flexmeasures/utils/plugin_utils.py

```diff
@@ -0,0 +1,131 @@
+import importlib.util
+import os
+import sys
+import warnings
+from importlib.abc import Loader
+
+import sentry_sdk
+from flask import Flask, Blueprint
+
+
+class ConfigurationWarning(Warning):
```

Comment:
I was thinking that these might be useful in all of FlexMeasures and looked up how FlexMeasures does it.

Have you seen `config_utils.are_required_settings_complete` and `config_utils.get_config_warnings`? What I like there is that we can add warnings in the default config, and (more crucial) we can tell developers where/how to add these settings. 

Maybe this can be done in a unified way (e.g. by adding an error handler for `ConfigurationError`, I'm not completely sure.

Reply:
I see. So we already made `required` and `warnable` categories for FlexMeasures config settings, with better support.

My goal was to have a place where plugins can register all of their config settings. I actually see 3 categories of config settings, corresponding with the logger levels `"error"`, `"warning"` and `"info"`. An example of the latter category being a config setting that overrides a default color, which is not really worth a warning.

Starting from the user/developer perspective, does any of the following options have your preference?

More verbose, but more easily extensible (see the warning `"message"` in the example below):
```
__settings__ = [
    {
        "name": "MY_PLUGIN_URL",
        "description": "URL used by my plugin for x.",
        "level": "error",
    },
    {
        "name": "MY_PLUGIN_TOKEN",
        "description": "Token used by my plugin for y.",
        "level": "warning",
        "message": "Without this token, my plugin will not do y.",
    },
    {
        "name": "MY_PLUGIN_COLOR",
        "description": "Color used to override the default plugin color.",
        "level": "info",
    },
]
```
Or shorter, but less easily changed:
```
__required_settings__ = {
    "MY_PLUGIN_URL": "URL used by my plugin for x.",
}
__warnable_settings__ = {
    "MY_PLUGIN_TOKEN": "Token used by my plugin for y.",
}
__facultative_settings__ = {
    "MY_PLUGIN_COLOR": "Color used to override the default plugin color.",
}
```

Reply:
I feel we should go to the first, more verbose version.
I would add an optional `parse_as` option, so FM can make sure the type fits.

Within FM, configs are growing up, so I propose they get their own `config` sub-package with

- __init__.py
- Readme.md
- utils.py
- defaults.py
- requirements.py

The latter would be where we put the requirements we talk about here. I call it requirements, as I don't think we should list all possible settings there. Or what do you think we should do?

For later, I can even imagine a post-config-loaded hook so plugins can control higher-level things if they want. E.g. check if at least one of two possible settings is there.

Reply:
I implemented a slightly modified version of the suggestion you chose, which is slightly less verbose but remains extensible. I also added checks for the `parse_as` option. Shall I open up an issue for the `config` sub-package, so we can continue that discussion there?

Reply:
Thanks. I like the structure you went with.

The more important part left out here (which would be the new ticket) is to make FlexMeasures use the code you wrote not only for plugin settings, but also for its own settings (thus encoding requirements and replacing `config_utils.are_required_settings_complete` and `config_utils.get_config_warnings`. The refactoring into a sub-package is secondary but also to be discussed there.

Reply:
I opened up issue https://github.com/SeitaBV/flexmeasures/issues/236. Not sure if you also want to mention the post-config-loaded hook idea there?